rust/macros/net.rs

```diff
@@ -0,0 +1,147 @@
+
+use super::helpers::*;
+
+pub fn rtnl_link_ops(ts: TokenStream) -> TokenStream {
+    let mut it = ts.into_iter();
+    let literals = &["maxtype", "policy", "slave_maxtype", "slave_policy"];
+
+    let mut found_idents = Vec::new();
+
+    let kind = get_byte_string(&mut it, "kind");
+    let netdevice = get_ident(&mut it, "type");
+
+    let mut callbacks = String::new();
+    let mut fields = String::new();
+
+    loop {
+        let name = match it.next() {
+            Some(TokenTree::Ident(ident)) => ident.to_string(),
+            Some(_) => panic!("Expected Ident or End"),
+            None => break,
+        };
+
+        assert_eq!(expect_punct(&mut it), ':');
+
+        if literals.contains(&name.as_str()) {
+            let literal = expect_literal(&mut it);
+            fields.push_str(&format!(
+                "{name}: {literal},\n",
+                name = name,
+                literal = literal
+            ));
+        } else {
+            let func = expect_ident(&mut it);
+            callbacks.push_str(&build_rtnl_links_callback(&name, &netdevice, &func, &kind));
+            found_idents.push(name);
+        }
+
+        assert_eq!(expect_punct(&mut it), ',');
+    }
+    expect_end(&mut it);
+
+    let callback_fields = found_idents
+        .iter()
+        .map(|name| format!("{}: Some(__rtnl_link_{}_callback_{}),", name, name, kind))
+        .collect::<Vec<String>>()
+        .join("\n");
+
+    let ops_struct = format!(
+        r#"
+             #[doc(hidden)]
+             #[used]
+             #[no_mangle]
+             pub static mut {kind}_LINK_OPS: kernel::net::rtnl::RtnlLinkOps = kernel::net::rtnl::RtnlLinkOps(kernel::bindings::rtnl_link_ops {{
+                 priv_size: core::mem::size_of::<<{netdevice} as kernel::net::device::NetDeviceAdapter>::Inner>(),
+                 kind: b"{kind}\0".as_ptr() as *const i8,
+                 {callback_fields}
+                 {fields}
+                 ..kernel::net::rtnl::RTNL_LINK_OPS_EMPTY                
+             }});
+         "#,
+        kind = kind,
+        netdevice = netdevice,
+        callback_fields = callback_fields,
+        fields = fields,
+    );
+
+    format!(
+        r#"
+         {}
+         // TODO: add #[link_section = ".data.read_mosty"] if x86
+         #[cfg(any(CONFIG_X86, CONFIG_SPARC64))]
+         #[link_section = ".data.read_mostly"]
+         {}
+         #[cfg(not(any(CONFIG_X86, CONFIG_SPARC64)))]
+         {}
+         "#,
+        callbacks, ops_struct, ops_struct,
+    )
+    .parse()
+    .expect("Error parsing formatted string into token stream.")
+}
+
+struct RtnlLinkValues {
+    callback_params: String,
+    return_type: String,
+    wrapper_before: String,
+    wrapper_after: String,
+    params: String,
+}
+
+impl RtnlLinkValues {
+    fn new(callback_params: &str, wrapper_before: &str, params: &str) -> Self {
+        Self {
+            callback_params: callback_params.to_string(),
```

Comment:
`to_owned` is preferred to `to_string`. `ToString::to_string` is generic string conversion function that is implemented for all types that implements `Display`, while `to_owned` is more clear in its intention that you want a owned copy.